tests/mesh/mesh_assign.C

```diff
@@ -0,0 +1,230 @@
+
+       // The boundary ids of the boundaries where the value is to be set.
+       const boundary_id_type all_bdry_ids[4] = {0, 1, 2, 3};
+       const std::set<boundary_id_type> bottom_right_bdry_ids(all_bdry_ids, all_bdry_ids + 1);
+
+       // The variables for which we are defining the bc.
+       std::vector<unsigned int> variables;
+       variables.push_back(0);
+
+       DirichletBoundary dirichlet_bc(bottom_right_bdry_ids, variables, &one);
+
+       system.get_dof_map().add_dirichlet_boundary(dirichlet_bc);
+
+       es.init();
+
+       if (mesh_type.compare("Mesh") == 0)
+       {
+        mesh_two = std::make_shared<Mesh>(*TestCommWorld);
+       }
+       else if (mesh_type.compare("replicated") == 0)
+       {
+        mesh_two = std::make_shared<ReplicatedMesh>(*TestCommWorld);
+       }
+       else if (mesh_type.compare("distributed") == 0)
+       {
+        mesh_two = std::make_shared<DistributedMesh>(*TestCommWorld);
+       }
+       else
+       {
+        libmesh_error_msg("Error: specified mesh_type not understood");
+       }
+
+       if(mesh_two_case_type.compare("from_memory") == 0)
+       {
+         /**
+          * Build another 2d 2x2 square mesh (mesh_two) covering [0.0, 1.0] x [0.0, 1.0]
+          * with linear Quad elements. We will later refine this mesh and replace
+          * mesh_one with it.
+          */
+
+          MeshTools::Generation::build_square(*mesh_two,
+                                              2, 2,
+                                              0., 1.,
+                                              0., 1.,
+                                              QUAD4);
+
+          // Uniformly refine mesh_two
+          std::unique_ptr<MeshRefinement> mesh_refinement(new MeshRefinement(*mesh_two));
+
+          mesh_refinement->uniformly_refine(1);
+          std::cout<<"Info from mesh two, from memory."<<std::endl<<std::endl;
+          mesh_two->print_info();
+
+          // Move mesh_two into mesh_one
+          system.get_mesh().clear();
+          system.get_mesh().assign(std::move(*mesh_two));
+          mesh_two->clear();
+          std::cout<<"Info from mesh one, after moving."<<std::endl<<std::endl;
+          system.get_mesh().print_info();
+
+          // Assert that the moved into mesh has the right number of elements.
+          CPPUNIT_ASSERT_EQUAL(system.get_mesh().n_elem(), dof_id_type(20));
+
+          // Reinit the dofs and other system related data structures
+          // based on its mesh
+          system.get_equation_systems().reinit_mesh();
+
+          // Reinit
+          system.get_equation_systems().reinit();
+
+          // Write out the mesh xda, this currently fails with DistributedMesh,
+          // running on more than 8 processors, when the mesh to be moved in is
+          // read from a file.
+          std::cout<<"Write out the mesh, read from memory."<<std::endl<<std::endl;
+          system.get_mesh().write("mesh.out.memory.moved.xda");
```

Comment:
If you are writing a mesh from a unit test, you will need to clean it up, otherwise we will get `make distclean` errors. This file should be added to the `CLEANFILES` list in Makefile.am in the unit tests directory.